moon/zmqpool.py

```diff
@@ -0,0 +1,105 @@
+"""
+  ZeroMQ Thread Pool for handling multiple ROS services
+"""
+
+from threading import Thread
+
+import zmq
+
+from osgar.bus import BusShutdownException
+
+
+class ZMQPool:
+    def __init__(self, config, bus):
+        bus.register('raw:gz' if config.get('save_data', False) else 'raw:null', 'response', 'timeout')
+        mode = 'REQ'
+        self.endpoint = config['endpoint']
+        self.timeout = config.get('timeout', 1)  # default recv timeout 1s
+        self.pool_size = config.get('pool_size', 5)
+
+        # support only mode='REQ'
+        self.thread = Thread(target=self.run_reqrep)
+
+        self.thread.name = bus.name
+        self.bus = bus
+
+    def start(self):
+        self.thread.start()
+
+    def join(self, timeout=None):
+        self.thread.join(timeout=timeout)
+
+    class ReqRepWorker(Thread):
+        def __init__(self, context, bus, endpoint):
+            Thread.__init__ (self)
+            self.context = context
+            self.bus = bus
+
+            self.rossocket = context.socket(zmq.REQ)
+            self.rossocket.RCVTIMEO = int(2000)
+            self.rossocket.connect(endpoint)
+            self.stop_requested = False
+
+        def stop(self):
+            self.stop_requested = True
+
+        def run(self):
+            worker = self.context.socket(zmq.REQ)
+            worker.setsockopt(zmq.LINGER, 0)
+            worker.RCVTIMEO = 1000 
+            worker.connect('inproc://reqrepbackend')
+
+            def run_loop():
+                while not self.stop_requested:
+                    try:
+                        worker.send(b"ready")
+                        while True:
+                            try:
+                                data = worker.recv()
+                                break
+                            except zmq.Again:
+                                if self.stop_requested:
+                                    return
+
+                        ident, msg = data.decode('ascii').split('|')
+                        self.rossocket.send_string(msg)
+                        rsp = self.rossocket.recv().decode('ascii')
```

Comment:
Now that this code performs the 'decode' function here, you need to remove it from all the places where 'send_request' is called.

Reply:
good point, thanks!